ndc-lambda-sdk/src/inference.ts

```diff
@@ -612,134 +660,365 @@ function unwrapNullableType(tsType: ts.Type, typeChecker: ts.TypeChecker): [ts.T
+ * NDC type name at the end, ensuring no two unique types have the same name (see 'CustomTypeNameRegistry').
+ */
+function makeUniqueTypeIdentifier(tsType: ts.Type, typeSource: TypeSource, typeChecker: ts.TypeChecker): UniqueTypeIdentifier {
+  // 'TypeId' is used to encode a string that represents the type identity.
+  // A string is necessary because we use it in map lookups and JS doesn't support using
+  // deep equals when objects are used as keys.
+  type TypeId = TypeWithDeclarationId | IntrinsicTypeId | LiteralTypeId | AnonymousObjectTypeId | AnonymousUnionTypeId | AnonymousIntersectionTypeId;
+  type TypeWithDeclarationId = { t: "d", f: string, s: number, ta: TypeId[] }
+  type IntrinsicTypeId = { t: "i", i: string }
+  type LiteralTypeId = { t: "l-n", v: number } | { t: "l-s", v: string } | { t: "l-bi", v: ts.PseudoBigInt }
+  type AnonymousObjectTypeId = { t: "o", p: [string, TypeId][] }
+  type AnonymousUnionTypeId = { t: "u", u: TypeId[] }
+  type AnonymousIntersectionTypeId = { t: "in", in: TypeId[] }
+
+  const compareList = <T>(compare: (a: T, b: T) => number, listA: T[], listB: T[]): number => {
+    const lengthCompare = listA.length - listB.length;
+    if (lengthCompare !== 0) return lengthCompare;
+    for (let i = 0; i < listA.length; i++) {
+      const itemCompare = compare(listA[i]!, listB[i]!);
+      if (itemCompare !== 0) return itemCompare;
+    }
+    return 0;
+  }
 
-    // If the type is present in the entrypoint, don't qualify the name
-    // If it is under the entrypoint's directory qualify with the subpath
-    // Otherwise, use the minimum ancestor of the type's location to ensure non-conflict
-    if (functionsFilePath === where) {
-      return nameOrGeneratedName;
-    } else if (short.length < where.length) {
-      return `${gqlName(short)}_${nameOrGeneratedName}`;
+  const compareTypeId = (typeIdA: TypeId, typeIdB: TypeId): number => {
+    if (typeIdA.t === "d" && typeIdB.t == "d") {
+      return typeIdA.f.localeCompare(typeIdB.f)
+        || typeIdA.s - typeIdB.s
+        || compareList(compareTypeId, typeIdA.ta, typeIdB.ta);
+    } else if (typeIdA.t === "i" && typeIdB.t == "i") {
+      return typeIdA.i.localeCompare(typeIdB.i);
+    } else if (typeIdA.t === "l-n" && typeIdB.t == "l-n") {
+      return typeIdA.v - typeIdB.v;
+    } else if (typeIdA.t === "l-s" && typeIdB.t == "l-s") {
+      return typeIdA.v.localeCompare(typeIdB.v);
+    } else if (typeIdA.t === "l-bi" && typeIdB.t == "l-bi") {
+      const aStr = `${typeIdA.v.negative ? "-" : ""}${typeIdA.v.base10Value}`;
+      const bStr = `${typeIdB.v.negative ? "-" : ""}${typeIdB.v.base10Value}`;
+      return aStr.localeCompare(bStr);
+    } else if (typeIdA.t === "o" && typeIdB.t == "o") {
+      return compareList(
+        ([nameA, pTypeIdA], [nameB, pTypeIdB]) => nameA.localeCompare(nameB) || compareTypeId(pTypeIdA, pTypeIdB),
+        typeIdA.p,
+        typeIdB.p
+      );
```

Comment:
I guess these strings could be constants somewhere?

Reply:
Eh, they're encoded into the type as literals; you can't use any other strings. Not really worth making them constants IMO. And they're local to this function.